### PR TITLE
RMT-1 voltmeter

### PR DESCRIPTION
**Pull request by sajidanower23**

### CODE REVIEW

File: src/libs/adc/VREF [hunk 1]
```diff
@@ -0,0 +1,2 @@
+VREFA+ => 3V from a precision voltage reference IC
+VREFA- => 0V from ground
```
Comment (hjed) [inline src/libs/adc/VREF]: it might be good to put these in a docs folder

File: src/libs/adc/adc.c [hunk 2]
```diff
@@ -0,0 +1,180 @@
+//------------------------------INCLUDES------------------------------//
```
Comment (hjed) [inline src/libs/adc/adc.c]: license header

File: src/libs/adc/adc.c [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    ADCIntClear(ADC0_BASE, sequence_num);
+    ADCSequenceDataGet(ADC0_BASE, sequence_num, adc_buffer);
+    /* call the callback */
+    adc_callback();
+}
+
+enum adc_return adc_init_pins(enum adc_pin *pins, uint8_t num_pins, bool interrupt_mode) {
+//        adc_assert(status == PRE_INIT);
+//        adc_assert(num_pins < 9);
+//        adc_assert(pins != (void *)0);
```
Comment (hjed) [inline src/libs/adc/adc.c]: commented code

File: src/libs/adc/adc.c [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+//------------------------------GLOBALS-------------------------------//
+int status = PRE_INIT;
+uint8_t active_pins = 0;
+uint32_t *adc_buffer = (void *)0;
+void (*adc_callback)(void) = (void *)0;
+
+
+//--------------------------LOCAL_FUNCTIONS---------------------------//
+/* If a module is not already enabled: enables it, then waits until it
```
Comment (hjed) [inline src/libs/adc/adc.c]: can you use `/**`  for these so it generates docs

File: src/libs/adc/adc.c [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        }
+        /* ready for capture */
+        status = POST_INIT;
+        return ADC_SUCCESS;
+}
+
+enum adc_return adc_capture_interrupt(uint32_t *buffer, void (*callback)(void)) {
+//    adc_assert(status == POST_INIT);
+//    adc_assert(buffer != (void *)0);
+//    adc_assert(callback !=(void *)0);
```
Comment (hjed) [inline src/libs/adc/adc.c]: commented code

File: src/libs/adc/adc.c [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    adc_buffer = buffer;
+    adc_callback = callback;
+    ADCProcessorTrigger(ADC0_BASE, sequence_num);
+    return ADC_SUCCESS;
+}
+
+
+
+uint32_t adc_capture_polling(uint32_t *buffer) {
+//    adc_assert(status == POST_INIT);
```
Comment (hjed) [inline src/libs/adc/adc.c]: commented code

File: src/libs/adc/adc.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
+/* Author: William Miles
```
Comment (hjed) [inline src/libs/adc/adc.h]: please use `/**` and doxygen style if possible

please include the license header

File: src/libs/adc/adc.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+enum adc_return adc_interrupt_disable();
+
+enum adc_return adc_interrupt_enable();
+
+#ifdef __cplusplus
+}
+#endif
+
+
```
Comment (hjed) [inline src/libs/adc/adc.h]: please use /** for all of these

File: src/libs/ros_echronos/ros_echronos.cmake [hunk 9]
```diff
@@ -16,25 +16,22 @@ FILE(GLOB ros_files ${ROS_ECHRONOS_DIR}/arch/tm4c/*.cpp ${ROS_ECHRONOS_DIR}/*.cp
...
+#            ${ros_include_files}
+#
+#    )
+#    add_custom_target(
+#            ros-echronos-doc-gen ALL
+#            COMMAND doxygen Doxyfile.ros-echronos-docs
+#            COMMAND cp -r html ../../../docs/doxygen
+#            DEPENDS ros-echronos-docs
+#    )
+#endif()
```
Comment (hjed) [inline src/libs/ros_echronos/ros_echronos.cmake]: please don't remove this it generates docs. You just need to install `doxygen`

File: src/modules/adc_test/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+include_directories(
+        ${CMAKE_INCLUDE_PATH}
+        ./
+        ${catkin_INCLUDE_DIRS}
+)
+
+set(BUILD_TOOLS_DIR ../../../build-tools)
+include(${BUILD_TOOLS_DIR}/module_template.cmake)
+
+#set(CPP_FILES adc_test.cpp can_wait_task.cpp)
```
Comment (hjed) [inline src/modules/adc_test/CMakeLists.txt]: does this need to be commented?

File: src/modules/adc_test/adc_test.cpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
+#include "boilerplate.h"
```
Comment (hjed) [inline src/modules/adc_test/adc_test.cpp]: header comment

File: src/modules/adc_test/adc_test.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    //enum adc_pin pins[NUM_PINS] = {AIN0, AIN1, AIN2, AIN3, AIN4, AIN5, AIN6, AIN7};
+    enum adc_pin pins[NUM_PINS] = {AIN0,AIN1,AIN2,AIN3};
+    adc_init_pins(pins, NUM_PINS, true);
+    while(1) {
+        if (capture) {
+            capture = false;
+            adc_capture_interrupt(out, adc_callback_fn);
+        }
+        //adc_capture_polling(out);
+        //UARTprintf("out = %d %d %d %d\n", out[0], out[1], out[2], out[3]);
```
Comment (hjed) [inline src/modules/adc_test/adc_test.cpp]: commented code (use a `#define` for debug code)

File: src/modules/voltmeter/voltmeter.cpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
+#include <rtos-kochab.h>
```
Comment (hjed) [inline src/modules/voltmeter/voltmeter.cpp]: header comment

File: src/modules/voltmeter/voltmeter.cpp [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    // this creates a node handle
+    ros_echronos::ROS_INFO("Entered CAN task. Initializing...\n");
+    ros_echronos::NodeHandle nh;
+    nh.init("voltmeter_fn", "voltmeter_fn", RTOS_INTERRUPT_EVENT_ID_CAN_RECEIVE_EVENT, RTOS_SIGNAL_ID_CAN_RECEIVE_SIGNAL);
+    ros_echronos::ROS_INFO("Done init\n");
+    nh_ptr = &nh;
+
+    // publisher
+    ros_echronos::ROS_INFO("pub init\n");
+    ros_echronos::Publisher<std_msgs::Float64> _pub("voltmeter", (std_msgs::Float64*)vm_buffer, 5, false);
```
Comment (hjed) [inline src/modules/voltmeter/voltmeter.cpp]: probably should be `/voltmeter` also I `_pub` is a bit of a weird naming convertion

Comment (sajidanower23) [reply src/modules/voltmeter/voltmeter.cpp]: Changed that to `/sensors/voltmeter` as that's already being used in the GUI

**Pull request merged**